File: mobile/src/lib/api/garmin.ts
import { apiClient } from '@/lib/api/client';
import type { PaceRange, PlanBlock } from '@/lib/api/plans';

export type GarminConnectResult = {
  status: 'connected' | 'needs_mfa';
  mfa_token: string | null;
};

export type GarminSyncResult = {
  status: 'done' | 'failed' | 'skipped';
  activities_synced: number | null;
  error_message: string | null;
};

/**
 * Garmin Connect has no public OAuth app flow (see garmin-sync skill): the
 * user's own Garmin credentials are submitted once, the backend logs in via
 * python-garminconnect and stores only the resulting session tokens.
 *
 * Accounts with 2FA enabled make this a two-step flow: this call returns
 * status "needs_mfa" with a short-lived mfa_token instead of connecting
 * directly; completeGarminMfa finishes it with the code Garmin emails.
 */
export function connectGarmin(garminEmail: string, garminPassword: string) {
  return apiClient.post<GarminConnectResult>('/api/v1/garmin/connect', {
    garmin_email: garminEmail,
    garmin_password: garminPassword,
  });
}

export function completeGarminMfa(mfaToken: string, mfaCode: string) {
  return apiClient.post<GarminConnectResult>('/api/v1/garmin/connect/mfa', {
    mfa_token: mfaToken,
    mfa_code: mfaCode,
  });
}

/**
 * Runs a Garmin activity sync and resolves only when it's finished (the
 * backend does the work synchronously, on a free host where a background
 * task could be killed mid-run). Throttled server-side, so a call within
 * the cooldown resolves quickly with status "skipped".
 */
export function syncGarmin() {
  return apiClient.post<GarminSyncResult>('/api/v1/garmin/sync');
}

export type WorkoutPushResult = {
  status: 'done' | 'failed';
  workout_name: string | null;
  error_message: string | null;
};

export type WorkoutPushPayload = {
  session_type: string;
  duration_min: number;
  structure: PlanBlock[];
  pace_range: PaceRange | null;
  hr_zone: number | null;
  rationale?: string | null;
  week_number?: number | null;
};

/**
 * Upload a plan session to Garmin Connect as a structured running workout. It
 * lands in the user's Garmin workout library and syncs to the paired watch on
 * Garmin's next connection (no direct device push exists).
 */
export function pushWorkoutToWatch(payload: WorkoutPushPayload) {
  return apiClient.post<WorkoutPushResult>('/api/v1/garmin/workout', payload);
}

File: backend/app/models/garmin.py
from datetime import datetime

from pydantic import BaseModel

from app.models.plan import Block, PaceRange


class GarminConnectRequest(BaseModel):
    garmin_email: str
    garmin_password: str


class GarminConnectResponse(BaseModel):
    status: str  # "connected" | "needs_mfa"
    mfa_token: str | None = None


class GarminMfaRequest(BaseModel):
    mfa_token: str
    mfa_code: str


class GarminSyncResponse(BaseModel):
    status: str  # "done" | "failed" | "skipped"
    activities_synced: int | None = None
    error_message: str | None = None


class WorkoutPushRequest(BaseModel):
    """A plan session to convert into a Garmin structured workout and upload
    to Garmin Connect (which then syncs it to the watch)."""

    session_type: str
    duration_min: int
    structure: list[Block] = []
    pace_range: PaceRange | None = None
    hr_zone: int | None = None
    rationale: str | None = None
    week_number: int | None = None


class WorkoutPushResponse(BaseModel):
    status: str  # "done" | "failed"
    workout_name: str | None = None
    error_message: str | None = None


class GarminCredentialsDocument(BaseModel):
    """Shape of a `garmin_credentials` collection document.

    The user's Garmin password is never persisted — only the encrypted,
    serialized Garmin session (see app/core/crypto.py).
    """

    user_id: str
    encrypted_tokens: str
    needs_relogin: bool = False
    connected_at: datetime
